docs/architecture-decision-records/0020-aggregation-for-tdr-preingest.md

```diff
@@ -0,0 +1,40 @@
+# 20. Aggregation for TDR Preingest
+
+**Date:** 2024-08-14
+
+## Context
+
+The first source to ingest into DR2 was parsed court documents transferred via TDR and parsed by TRE. The upstream workflow for this type of record enforces a 1-1 relationship between a transfer consignment and a record. We're now ready to create a preingest route for standard TDR transfers.
+
+Standard TDR transfers contain multiple files within a consignment; the count and size of the files is at the discretion of the transferring user. While the Preservation System ingest process is most efficient when ingesting batches, as opposed to individual files, the size of the batches are unlikely to match the user decisions upstream of DR2; we want to design a preingest process for TDR to be most efficient for the Preservation System.
+
+## Decision
+
+We will implement an aggregation process as part of our TDR Preingest workflow, we will request TDR notify us of new records for ingest with a message per record to allow us to aggregate. Our aggregation process will use the same AWS services as our ingest process - SQS, Lambda, Step Functions, and DynamoDB.
+
+### Aggregation Design
+
+![TDR Aggregation Diagram](/docs/images/adr/0020/tdr-aggregation-diagram.png)
+
+We will extend our usage of the existing dr2-ingest-lock table (created for [`0016-generic-ingest-idempotency.md`](./0016-generic-ingest-idempotency.md)) to act as a lock table within our aggregation. We will add a `groupId` attribute and GSI on this. And a `message` attribute to store the JSON string of the message received from TDR.
+
+Messages will be received into an SQS queue with an Event Source Mapping configured to a Lambda function. The Lambda will attempt to write each message to the lock table. The Event Source Mapping will be set with a `MaximumBatchingWindowInSeconds` of 300 seconds and `BatchSize` of 10,000; the Lambda will only be invoked when the SQS poller has 10,000 messages ready, 300 seconds have elapsed since the first message was received by the poller, or the Lambda payload is 6MB. To enable processing of batches larger than 10,000 records we will cache the current `groupId` within the Lambda container and delay processing of the group for a configurable period whilst other messages can still be added. Once this period has elapsed, we will collect the messages from the lock table and use these to build a DR2 BagIt-like ingest package. The items in the lock table will be deleted as normal when the ingest process is complete.
+
+As we're using a GSI within this solution we will add a step to the Generic Ingest workflow to check all items have been removed from the lock table and retry the "collect and build" process if any items remain. To aid terminology, we will **aggregate into groups and ingest in batches**, where a batch is a combination of `groupId` and `retryCount`.

```

Comment:
1. Missing comma after "within this solution"
2. You use the term "collect and build" in quotes but you haven't used it before

Reply:
2. Err, I don't know how to describe it! I say `Once this period has elapsed, we will collect the messages from the lock table and use these to build a DR2 BagIt-like ingest package.` in the paragaph above, any thoughts?

Reply:
Ohhh, you're referring to the process directly above; hmm, maybe "retry **this** "collect and build" process"? Also, do we need a new paragraph? In fact, could the new paragraph start after the "...Lambda payload is 6MB." sentence, instead?

Reply:
Changes paragraph split, and reworked to
> As we're using a GSI within this solution, we will add a step to the Generic Ingest workflow to check all items have been removed from the lock table. If any items remain, we will run repeat the package building process; creating a new DR2 BagIt-like package for the remaining messages.